ZEP9 (phase 1): add clarifications for extension naming (#330)
zarr-developers/zarr-specs: docs/v3/core/index.rst

Therefore, data is only marked with the respective major version, unknown
features are auto-discovered via the metadata document.

Notably, this excludes extensions such as codecs, data types, chunk grids
and storage transformers from the compatibility of the core specification, as
well as store support. However, extensions and stores are also RECOMMENDED to 

> follow this stability policy.

Reviewer: this applies to the original text, but as long as we are editing it: what exactly does this statement mean? Why are we talking about versioning stores here? And why do we care if extensions use a specific versioning scheme? I think a motivation is necessary, because otherwise it's not clear.

    contain. For example, the 32-bit signed integer data type defines binary
    representations for all integers in the range −2,147,483,648 to
    2,147,483,647. This specification only defines a limited set of data types,

>     but :ref:`extensions<extensions_section>` may define other data types.

Reviewer: ```suggestion
    but additional data types can be defined as :ref:`extensions<extensions_section>`.
```

Reviewer: actually, as this is a field definition, I don't think we should say anything about extensions

    The chunks_ of an array_ are organised into a grid. This
    specification only considers the case where all chunks_ have the
    same chunk shape and the chunks form a regular grid. However,

>     :ref:`extensions<extensions_section>` may define other grid types such as
>     rectilinear grids.

Reviewer: ```suggestion
    additional chunk grids can be defined as :ref:`extensions<extensions_section>` 
```

.. image:: terminology-read.excalidraw.png
  :width: 600

*Extension points*

>     Locations within a `metadata document_` where extension-related

Reviewer: ```suggestion
    Fields in a `metadata document_` where extension-related
```

> *Extension points*
> 
>     Locations within a `metadata document_` where extension-related
>     metadata can be found. Current extension points are listed in the core spec,
>     e.g. `codecs`, `data_type`. See :ref:`extension points <extensions_section>` below.

Reviewer: ```suggestion
*Extension points*

    Fields in a `metadata document_` that can be extended to allow values not defined in this specification. See :ref:`extension points <extensions_section>` below.
```

*Extensions*

    Components defined in a `metadata document`_ to
    configure how metadata are interpreted by implementations. These
    components include codecs, data types, chunk key encodings, chunk grids and

>     storage transformers. See :ref:`extension points <extensions_section>` below.

Reviewer: what is the difference between the "Extension points" and "Extensions" sections? It seems like we might only need one.

Author: This is trying to convey a class/instance style relationship:

| **Places to put things** | **Things to put there** |
|----------------------|---------------------|
| e.g. the codecs field | a particular codec |
| extension point      | extension           |
| type of extension    | extension           |
| extension            | ...what here? ...   |

Reviewer: I don't think we should model "extension" as a class? Extensibility is property of a set of values. A set of values is extensible if new elements can be added. A set of values is not extensible if no new values can be added. We don't add anything to these definitions by thinking of "extension" as a class that has implementations / instances.

Reviewer: I don't think we need two terms -- the definition of "extension point" is sufficient to convey "a field with a non-static set of values".

*Core*

>     Core indicates a feature or concepts defined within the Zarr v3
>     specification as defined in this repository. Note, however, that certain
>     core features are explicitly marked as optional for implementations.

Reviewer: ```suggestion
    Core metadata is that which is defined in the Zarr v3 specification.
```

Author: I think "core" is used in more than just metadata (other than that basically everything in zarr.json is metadata). The more important part of this is that core != MUST.

Reviewer: Why do we need to make a negative statement about what core is not? It should be sufficient to say what it is.

Author: Because there is some confusion on this topic. I'll push another formulation.

mandatory names:

.. _array-metadata-zarr-format:

``zarr_format``
""""""""""""""""

    An integer defining the version of the storage specification to which the
    array store adheres, must be ``3`` here.

.. _array-metadata-node-type:

``node_type``
"""""""""""""""

    A string defining the type of hierarchy node element, must be ``array``
    here.

.. _array-metadata-shape:

``shape``
"""""""""

    An array of integers providing the length of each dimension of the
    Zarr array. For example, a value ``[10, 20]`` indicates a
    two-dimensional Zarr array, where the first dimension has length
    10 and the second dimension has length 20.

.. _array-metadata-data-type:

``data_type``
"""""""""""""

    The data type of the Zarr array.

    ``data_type`` is an :ref:`extension point<extensions_section>`
    and MUST conform to the :ref:`extension-definition`.

    If the data type is defined in
    this specification, then the value must be the data type
    identifier provided as a string. For example, ``"float64"`` for
    little-endian 64-bit floating point number.

.. _array-metadata-chunk-grid:

``chunk_grid``
""""""""""""""

>     The chunk grid of the Zarr array.

Author: That probably means pushing more of the content in this section to the subpages. Agreed?

>     The chunk grid of the Zarr array.

Author: Leaving such aesthetics to a follow up.

--------------

Each Zarr array in a hierarchy must have an array metadata document, named
``zarr.json``.

Mandatory
^^^^^^^^^

This document must contain a single object with the following
mandatory names:

.. _array-metadata-zarr-format:

``zarr_format``
""""""""""""""""

>     An integer defining the version of the storage specification to which the
>     array store adheres, must be ``3`` here.

Reviewer: ```suggestion
    An integer denoting the version of the Zarr format. It must be the value ``3``.
```

.. _array-metadata-node-type:

``node_type``
"""""""""""""""

>     A string defining the type of hierarchy node element, must be ``array``
>     here.

Reviewer: ```suggestion
    A string denoting the type of hierarchy node element. It must be the value ``"array"``.
```

Author: I noted this in https://github.com/zarr-developers/zarr-specs/issues/329. Thanks.

.. _array-metadata-shape:

``shape``
"""""""""

    An array of integers providing the length of each dimension of the
    Zarr array. For example, a value ``[10, 20]`` indicates a
    two-dimensional Zarr array, where the first dimension has length
    10 and the second dimension has length 20.

.. _array-metadata-data-type:

``data_type``
"""""""""""""

    The data type of the Zarr array. If the data type is defined in
    this specification, then the value must be the data type
    identifier provided as a string. For example, ``"float64"`` for
    little-endian 64-bit floating point number.

    The ``data_type`` value is an :ref:`extension point<extensions_section>`
    and may be defined by a data
    type extension. If the data type is defined by an extension, then the value

>     may be either a plain string (incl. URI) or an object containing the members ``name``

Reviewer: ```suggestion
    may be either a string or an object containing the members ``name``
```
a URI _is_ a string, so I don't think we need to mention URIs here without a compelling reason.

Author: And if we give this particular type a string a name ("extension identifier"?) since it's defined as:

`(^[a-z0-9-_.]+$|^https?:\/\/[^/?#]+[^?#]*$)`

Reviewer: ah i didn't see that this is already in the context of data type extensions. in that case, we should be able to say that it must be a data type identifier, perhaps linking to the definition of that type.

    same dimension name across multiple arrays within the same Zarr hierarchy,
    but extensions or specific applications may do so.

.. _array-metadata-extensions:

Unknown
^^^^^^^

> All other keys found in the metadata object MUST be interpreted
> following the :ref:`Extensions section <extensions_section>`.

Reviewer 2: I think this is misleading here. Array metadata under 3.1 must not have additional keys. Observing additional keys would indicate another version. With our current proposals, extensions cannot add root-level metadata keys.

```suggestion
Other keys are reserved for future versions of the specification.
Therefore, there MUST NOT be other keys in the metadata object.
See :ref:`Extensions section <extensions_section>` for more details.
```

Reviewer 3: Wouldn't that be a regression? Array metadata under 3.0 with an unknown metadata field with ``"must_understand": false`` was valid, and now would be invalid.

Reviewer 2: Fair enough. `"must_understand": false` should still be allowed, but discouraged in favor of explicit extensions.

Reviewer 2: Fair enough. `"must_understand": false` should still be allowed, but discouraged in favor of explicit extensions.

Author: Added a short section locally. will push shortly. We may want to expand it.

    same dimension name across multiple arrays within the same Zarr hierarchy,
    but extensions or specific applications may do so.

.. _array-metadata-extensions:

> Extensions

Reviewer: "extension" is being used in two ways in this PR.
- externally-defined data types, codecs, etc
- new fields added to `zarr.json`

I think these two cases are _very_ different, and so different terminology should be used to distinguish them. I'd suggest describing fields like `data_type` as having extensible values, and maybe use the term "metadata extensions" to refer to adding new fields to `zarr.json`

^^^^^^^^^^

> All other names found in the metadata object MUST be interpreted

Reviewer: ```suggestion
All other fields found in the metadata object MUST be interpreted
```

Reviewer: or "keys", if that's what we are using

Group metadata
--------------

Mandatory
^^^^^^^^^

A Zarr group metadata object must contain the following mandatory key:

``zarr_format``
"""""""""""""""

    An integer defining the version of the storage specification to which the
    array store adheres, must be ``3`` here.

``node_type``
"""""""""""""""

    A string defining the type of hierarchy node element, must be ``group``
    here.

Optional
^^^^^^^^

Optional keys:

``attributes``
""""""""""""""

    The value must be an object. The object may contain any key/value
    pairs, where the key must be a string and the value can be an arbitrary
    JSON literal. Intended to allow storage of arbitrary user metadata.

Unknown
^^^^^^^

> All other keys found in the metadata object MUST be interpreted

Reviewer 2: Same comment as above.

Group metadata
--------------

Mandatory
^^^^^^^^^

A Zarr group metadata object must contain the following mandatory key:

``zarr_format``
"""""""""""""""

    An integer defining the version of the storage specification to which the
    array store adheres, must be ``3`` here.

``node_type``
"""""""""""""""

>     A string defining the type of hierarchy node element, must be ``group``

Reviewer: ```suggestion
    A string denoting the type of hierarchy node element. It must be the string ``"group"``
```

Author: I'm skipping comments which are not extension related and are outside of the changes proposed here. (Marked with 👀 to note that I've reviewed them)

    here.

Optional
^^^^^^^^

Optional keys:

``attributes``
""""""""""""""

    The value must be an object. The object may contain any key/value
    pairs, where the key must be a string and the value can be an arbitrary
    JSON literal. Intended to allow storage of arbitrary user metadata.

Extensions
^^^^^^^^^^

All other names found in the metadata object MUST be interpreted
following the `extensions_section`_.

> An implementation MUST fail to open Zarr hierarchies, groups
> or arrays if any metadata fields are present which (a) the
> implementation does not recognize and (b) are not explicitly
> set to ``"must_understand": false``.

Reviewer: ```suggestion
```
The normative content should be in the linked section

> This core specification defines a limited set of data types to

Reviewer: ```suggestion
This specification defines a limited set of data types to
```

represent boolean values, integers, and floating point
numbers. These can be found under :ref:`Data Types<data-types-list>`.

:ref:`Extensions<extensions_section>` may define additional

> data types. All of the data

Reviewer: this applies to the old text, but the additional data type _is_ the extension. An additional data type is not defined by an extension -- it's defined by a spec.

string. However, extensions may use any JSON value to identify a data
type.

> Additionally to these base types, an implementation should also handle the

Reviewer: ```suggestion
In addition to these base types, an implementation should also handle the
```

4. The chunk array ``A`` is equal to ``EC[0]``.

.. _codec-specification:

Core codecs
-----------

> This specification defines a set of codecs ("core codecs") which all Zarr implementations SHOULD implement in

Reviewer 2: Where is this list, actually?

Author: I interpret this as anything listed on https://zarr-specs--330.org.readthedocs.build/en/330/v3/codecs/index.html

Reviewer 2: I think in a previous draft we had just `bytes` and `transpose` as core codecs.

Author: Opened an issue: https://github.com/zarr-developers/zarr-specs/issues/341

Changes to the list of core codecs MUST be made via the same protocol used for
changing the Zarr v3 specification. Changes to the list of core codecs SHOULD be made
in close collaboration with extant Zarr v3 implementations. A new core codec SHOULD be added to the
list when a sufficient number of Zarr implementations support or intend to support that codec.
An existing core codec SHOULD be removed from the list when a sufficient number of implementation
developers and Zarr users deem the codec worth removing, e.g. because of a technical flaw in the
algorithm underlying the codec.

Extension codecs
----------------

To allow for flexibility to define and implement new codecs, the
list of codecs defined for an array MAY contain codecs which are
defined in separate specifications. In order to refer to codecs in array metadata
documents, each codec must have a unique identifier, which is either
a known "`raw name <extension-naming-raw-names>`_" or
a "`URL-based name <extension-naming-url-based-names>`_" as defined under :ref:`extensions_section`.
For ease of discovery, it is
recommended that codec specifications are contributed to the
registry of extensions
(`zarr-extensions`_).

> A codec specification must declare the codec identifier, and describe

Reviewer 2: ```suggestion
A codec specification MUST declare the codec identifier, and describe
```

(or cite documents that describe) the encoding and decoding algorithms
and the format of the encoded data.

> A codec may have configuration parameters which modify the behaviour

Reviewer 2: ```suggestion
A codec MAY have configuration parameters which modify the behaviour
```

4. The chunk array ``A`` is equal to ``EC[0]``.

.. _codec-specification:

Core codecs
-----------

> This spec defines a set of well-known codecs ("core codecs") which all Zarr implementations SHOULD implement in

Reviewer: ```suggestion
This specification defines a set of codecs ("core codecs") which all Zarr implementations SHOULD implement in
```

order to ensure a minimal level of interoperability between Zarr implementations.
The list of core codecs is part of the Zarr v3 specification.
Changes to the list of core codecs MUST be made via the same protocol used for
changing the Zarr v3 specification. Changes to the list of core codecs SHOULD be made
in close collaboration with extant Zarr v3 implementations. A new core codec SHOULD be added to the
list when a sufficient number of Zarr implementations support or intend to support that codec.
An existing core codec SHOULD be removed from the list when a sufficient number of implementation
developers and Zarr users deem the codec worth removing, e.g. because of a technical flaw in the
algorithm underlying the codec.

Extension codecs
----------------

To allow for flexibility to define and implement new codecs, the
list of codecs defined for an array MAY contain codecs which are
defined in separate specifications. In order to refer to codecs in array metadata
documents, each codec must have a unique identifier, which is either

> a known "raw name" or as a URI as defined under :ref:`extensions_section`.

Reviewer: ```suggestion
a known "raw name" or a URI as defined under :ref:`extensions_section`.
```

an object.

For example::

    {
        "name": "<name>",        # "raw name" or URL-based name
        "configuration": { ... } # optional
    }

.. _extension-definition-short-hand-name:

Short-hand names
^^^^^^^^^^^^^^^^

Instead of extension objects, short-hand names MAY be used if no
configuration metadata is required. They are equivalent to extension
objects with just a `name` key.

.. _extension-definition-must-understand:

`must_understand`
^^^^^^^^^^^^^^^^^

If such an object is present, the field `must_understand` is implicitly set to
`True` and an object MAY explicitly set `must_understand=False` if
implementations can ignore its presence.

An implementation MUST fail to open Zarr groups or arrays if any
metadata fields are present which (a) the
implementation does not recognize and (b) are not explicitly
set to ``"must_understand": false``.

`must_understand=False` is not supported for the following extension points:
data type, chunk grid, and chunk key encoding.

Use of `must_understand=False` to add top-level keys is discouraged in favor
of the explicit use of :ref:`extension-points`.

.. _extension-naming:

Extension naming
----------------

> The `name` field of an extension can take two forms: **raw names** and **URL-based names**.

Reviewer: I feel like it's more clear to say that there are two types of extensions -- those that are centrally registered (these may have raw names OR namespaced names), and those that are not centrally registered (these may only have namespaced names).

> The `name` field of an extension can take two forms: **raw names** and **URL-based names**.

Reviewer 4: Why would we be registering namespaced codecs? In my head these were mutually exclusive categories. What you describe sounds more confusing and ambiguous.

> The `name` field of an extension can take two forms: **raw names** and **URL-based names**.

Reviewer: if codec `foo.codec` becomes popular enough that its owners want to register it centrally, doesn't it make sense to keep the same name?

> The `name` field of an extension can take two forms: **raw names** and **URL-based names**.

Reviewer: maybe it would help to sketch out the process for "publishing" an extension that was previously unpublished -- it sounds like you would want the prefix to be removed?

> The `name` field of an extension can take two forms: **raw names** and **URL-based names**.

Reviewer 5: In principle, someone else might be using the same namespace for their own private use.  Registering a namespaced name creates the possibility of a conflict.

> The `name` field of an extension can take two forms: **raw names** and **URL-based names**.

Reviewer: I think this is a manageable risk. People creating their own extensions in private should be aware that there can be no guarantee that the name they are using will be globally reserved for them. Forcing registered extensions to change their name seems disruptive and I would rather avoid that.

> The `name` field of an extension can take two forms: **raw names** and **URL-based names**.

Reviewer: but I think my main concern is conflating the structure of the extension name (raw vs prefixed) with its intended scope -- the text currently reads as if namespaced names cannot be used when broad interoperability is desired. But IMO organizations should be allowed to use namespaced names in this context, in the same way that geoarrow [defines their extensions under the `geoarrow` prefix.](https://geoarrow.org/extension-types.html#extension-names)

.. _extension-naming-raw-names:

Raw names
^^^^^^^^^

Raw names are centrally registered names which can be used without prefix.

> Raw names MUST be assigned within a central repository.

Reviewer 4: ```suggestion
Raw names MUST be assigned within a central repository, in order to ensure their uniqueness.
```

by only lower case letters a-z, numerals 0-9, underscores, dashes, and dots.
Raw name assignment is managed through the `zarr-extensions`_
Github repository, where extensions and their specification are listed.
The Zarr Steering Council or by delegation a
maintainer team reserves the right to refuse name assignment at its own
discretion.

- **Example:** ``zstd``
- **Accepted regex:** ``^[a-z][a-z0-9-_.]+$``

.. _extension-naming-url-based-names:

URL-based names
^^^^^^^^^^^^^^^

URL-based names delegate the assignment of unique identifiers to
the well-established addressing mechanism of the web.

URL-based names are decentralized and MAY be used by any extension without further coordination.
Entities defining a URL-based name SHOULD have appropriate
authority over the URL. A persistent redirecting URL like PURL MAY be used.
URLs have been chosen due to their potential for being self-documenting.
While a URL SHOULD resolve to a human-readable
explanation of the extension, preferably including a JSON schema definition
of the extension metadata, implementations are not expected to resolve
URLs during processing.

- **Example:** ``https://example.com/zarr3/consolidated-metadata``

> - **Accepted regex:** ``^https?:\/\/[^/?#]+[^?#]*$``

Reviewer: This limits the degree of nesting to 2, i.e. `a.b.c` would not be a permitted name under this proposal. Is that intended?

> - **Accepted regex:** ``^https?:\/\/[^/?#]+[^?#]*$``

Reviewer: also, do we want to reserve the prefix `zarr`?

Extension versioning
--------------------

Extensions with **raw names** SHOULD follow the
compatibility and versioning v3 `stability policy`_.

> For extensions with **URL-based names**, there are no guarantees in terms of

Reviewer 4: ```suggestion
For extensions with **namespaced names**, there are no guarantees in terms of
```

facilitates multiple implementations of an extension.

For extensions with raw names, the `zarr-extensions`_ repository
SHOULD either contain the specification or link to it.

For extensions with URL-based names, it is RECOMMENDED that the URL resolve to
a specification of the extension. Additionally, URL-based extensions MAY also register
themselves under the `zarr-extensions`_ repository for better discovery.

Because the ``fill_value`` metadata key is dependent on the data type, 
extension data types SHOULD specify permitted values for the ``fill_value`` in
their specification.

.. _extension-guidance:

Guidance for extension authors
------------------------------

*This section is non-normative and provides assistance for the authors of
extensions, especially those who are just getting started.*

* If you are just getting started, use the URL of your work-in-progress as an
  identifier for your extension. The GitHub link, including the branch if you
  would like, makes a fine choice. This says to the community that this is a
  draft, and if they are interested in the details, they can follow the URL to
  find out more.

> * When developing an extension for which you intend to register a short name,
>   you may wish to test it using the short name even before you have registered
>   it. However, you MUST register the name before using the extension for
>   non-test purposes/for purposes where interoperability with other
>   implementations/users is a concern.

Reviewer 4: ```suggestion
* If you intend to distribute data widely using your extension, you SHOULD register your extension using Raw name, rather than a namespaced name.
```

* If you are implementing a well-known extension like a data type or codec that
  is already referred to by name in the community, you may want to check the `zarr-extensions`_
  repository to see if someone has already implemented the extension.

* For raw names that are coming from well-known projects, use the same prefix followed
  by a dot for requesting your raw name, e.g. "numcodecs.". Other examples of prefixes can

>   be found in the `zarr-extensions`_ repository.

Reviewer 4: Note that these would be considered namespaced extensions under my proposal.

Reviewer: related to my [comment above](https://github.com/zarr-developers/zarr-specs/pull/330#discussion_r1983527438) but we should clarify early on that registered names can also be namespaced.

The `name` field of an extension is an identifier that has been registered
prior to release in any implementation within the `zarr-extensions`_ Github
repository, where extensions and their specification are listed. The Zarr
Steering Council or by delegation a maintainer team reserves the right to
refuse name assignment at its own discretion.

.. _extension-naming-registered-names:

Registered names consist of a single string that is unique within the Zarr ecosystem.
Registered names are intended for well-known extensions aimed at broad adoption and maximum interoperability.
Registered names are unique and immutable.

Registered names MUST start with one lower case letter a-z and then be followed
by only lower case letters a-z, numerals 0-9, underscores, dots and dashes.

- **Accepted regex:** ``^[a-z][a-z0-9-_.]+$``
- **Valid examples:**
    - ``zstd``
    - ``numcodecs.adler32``
- **Invalid examples:**
    - ``foo/bar``
    - ``foo:bar``

.. note::
  In previous versions of the v3 spec, the name of an extension was required
  to be a URI. That is now discouraged for new extensions, though, for
  backwards compatibility with existing extensions, URIs names are still

>   permitted.

Reviewer 6: Does this note need to be updated or perhaps removed to better reflect the current position?

Reviewer 6: I.e., IIUC, registered extensions will not have a URI name, and registration is encouraged. But for now we'd like to leave open the option that URIs can also be used as extension names to provide a decentralised mechanism to avoid extension name clashes. So saying "URIs are discouraged for new extensions" in this note is potentially confusing?

Author: I think this is fundamentally what over half the conversation here was about. I agree that the current state is confusing, or at least ambiguous, but that is the current state of the spec and I'd leave this to a follow on PR.

array       chunk grid              :ref:`chunk-grid <array-metadata-chunk-grid>`                      :ref:`chunk-grid-list`
array       chunk key encoding      :ref:`chunk-key-encoding <array-metadata-chunk-key-encoding>`      :ref:`chunk-key-encoding-list`
array       codecs                  :ref:`codecs <array-metadata-codecs>`                              :ref:`codec-list`
array       storage transformer     :ref:`storage-transformers <array-metadata-storage-transformers>`  :ref:`storage-transformer-list`
=========== ======================= ================================================================== ================================

Note, that ``fill_value`` is not its own extension point, but is dependent on the data type.

New extension points may be proposed to the Zarr community through the ZEP
process. See `ZEP 0 <https://zarr.dev/zeps/active/ZEP0000.html>`_ for more information.

.. _extension-definition:

Extension definition
--------------------

.. _extension-definition-object:

Objects
^^^^^^^

In `metadata documents`_, extensions can be encoded either as objects or as
short-hand names.

> If using an objection definition, the member ``name``

Reviewer 5: object

Raw names MUST be assigned within a central repository.
Raw names are unique and immutable.
Raw names MUST start with one lower case letter a-z and then be followed
by only lower case letters a-z, numerals 0-9, underscores, and dashes.
Raw names MUST NOT use a dot character `.`, to avoid confusion with namespaced extensions.

Raw name assignment is managed through the `zarr-extensions`_
Github repository, where extensions and their specification are listed.
The Zarr Steering Council or by delegation a
maintainer team reserves the right to refuse name assignment at its own
discretion.

- **Example:** ``zstd``
- **Accepted regex:** ``^[a-z][a-z0-9-_]+$``

.. _extension-naming-namespaced-names:

Namespaced names
^^^^^^^^^^^^^^^^

Namespaced names are intended for private extensions and for experimental and development purposes.
Namespaced names start with a prefix of one or more parts, each separated by the `.` character.

Namespaced names are not centrally managed and MAY be used by any extension without coordination.

- **Example:** ``myorg.my-private-extension``
- **Accepted regex:** ``^([a-z][a-z0-9-_]+\.)+[a-z][a-z0-9-_]+$``

URIs as names
-------------

> In an earlier draft of this spec, the name of an extension codec was required to be a URI that

Reviewer 5: "version" rather than draft

Author: Cheers. Changed in my recent push.

        "configuration": { ... } # optional object
    }

.. _extension-definition-short-hand-name:

Short-hand names
^^^^^^^^^^^^^^^^

Instead of extension objects, short-hand names MAY be used if no
configuration metadata is required. They are equivalent to extension
objects with just a `name` key.

.. _extension-definition-must-understand:

`must_understand`
^^^^^^^^^^^^^^^^^

An extension object is interpreted to have an implicit field `must_understand` set to
`True`, unless otherwise stated. An extension object MAY explicitly set `must_understand=False` if
implementations can ignore its presence.

An implementation MUST fail to open Zarr groups or arrays if any
metadata fields are present which (a) the
implementation does not recognize and (b) are not explicitly
set to ``"must_understand": false``.

`must_understand=False` is not supported for the following extension points:
data type, chunk grid, and chunk key encoding.

Use of `must_understand=False` to add top-level keys is discouraged in favor
of the explicit use of :ref:`extension-points`.

.. _extension-naming:

Extension naming
----------------

The `name` field of an extension is an identifier taking one of two forms:

> **registered names** (as simple strings) and **unregistered names** (as URIs).

Reviewer: i'm confused, going by [this comment](https://github.com/zarr-developers/zarr-specs/pull/330#issuecomment-2724936994) I thought we were dropping the requirement that unregistered names be URIs?

Reviewer: so to be clear, is it required that unregistered names be URIs?

Author: Short answer: Yes. But, I hadn't realized that it wasn't generally accepted that URIs need not be `http://host/stuff`. So what I've tried to make clear in the text is that there are also [`urn:`](https://en.wikipedia.org/wiki/Uniform_Resource_Name)s and that we're going to pick a short and easy to understand  prefix for any unregistered/namespaced name. (The [`tag:` scheme](https://en.wikipedia.org/wiki/Tag_URI_scheme) may be of interest as well.)

Implementations SHOULD be able to resolve multiple names to the same
implementation to support unregistered names which are subsequently registered.

.. _extension-naming-registered-names:

Registered names
^^^^^^^^^^^^^^^^

Registered names consist of a single string that is unique within the Zarr ecosystem, with no prefix.
Registered names are intended for well-known extensions aimed at broad adoption and maximum interoperability.
Registered names are unique and immutable.
Registered names MUST start with one lower case letter a-z and then be followed
by only lower case letters a-z, numerals 0-9, underscores, dots and dashes.

Registered names MUST be assigned within a central repository, `zarr-extensions`_
a Github repository, where extensions and their specification are listed.
The Zarr Steering Council or by delegation a
maintainer team reserves the right to refuse name assignment at its own
discretion.

- **Accepted regex:** ``^[a-z][a-z0-9-_.]+$``
- **Valid examples:**
  - ``zstd``
  - ``numcodecs.adler32``
- **Invalid examples:**

>   - ``foo/bar``

Reviewer 2: ```suggestion
  - ``foo/bar``
  - ``foo:bar``
```

.. _extension-naming-unregistered-names:

Unregistered names
^^^^^^^^^^^^^^^^^^

Unregistered names are intended for private extensions and for experimental and development purposes.

> Unregistered names are not centrally managed and MAY be used by any extension without coordination.

Reviewer: if these two statements are true, then why are we imposing any requirements on the structure of unregistered?

Unregistered names consist of URIs, which are prefixed with a scheme beginning
with a letter and followed by any number of letters, numbers, plus symbols,

> dashes or dots and then followed by a colon.

Reviewer: I think it would be worth explaining why total namespace separation between registered names and unregistered names is a design goal, and why URIs are a particularly suitable way to achieve that goal. 

I think it's significant that the arrow does not use URIs, and instead suggests that extension names be namespaced with a "." character. If we start from that position, and we are still worried about name collisions between registered extensions and unregistered extensions (and this needs to be discussed on a technical level, I think), then we could achieve namespace isolation by using a reserved namespace for registered extensions, or requiring that unregistered extensions use a specific namespace. URIs seem a lot more complicated than "."-separated strings, so I think URIs also need extra explanation / motivation, and that is not present in this PR right now.

> dashes or dots and then followed by a colon.

Reviewer 4: I believe @joshmoore should be the one to provide this justification.

- **Identifying regex:** ``^([a-z][a-z0-9-_]+\.)+[a-z][a-z0-9-_]+:``

URIs (`Uniform Resource Identifiers <https://en.wikipedia.org/wiki/Uniform_Resource_Identifier>`_)
are a well-known mechanism to identify resources on the internet and extension authors are

> encouraged to explore further documentation on which identifiers might best express their intent.

Reviewer: The language about resources on the internet here is confusing. I can create a zarr extension locally, on my own computer, without connecting to the internet. That suggests that zarr extensions are not really resources on the internet. Rather, zarr extensions are _code_, which might have a specification / source code on the internet, but those documents are not the same thing as the extension itself.

Reviewer 2: Stealing from the wikipedia article:
```suggestion
URIs (`Uniform Resource Identifiers <https://en.wikipedia.org/wiki/Uniform_Resource_Identifier>`_)
are a well-known mechanism to identify abstract or phyiscal resources and extension authors are
encouraged to explore further documentation on which identifiers might best express their intent.
```

Aware that not all extension developers will want to immediately register a name,
the goal of using URI identifiers is to provide a large and flexible namespace which
balances the needs of developers building new extensions with a extensible mechanism
which the Zarr community can make use of in the years to come.

URLs (`Uniform Resource Locators <https://en.wikipedia.org/wiki/URL>`_) are one class of URIs which
provide a mechanism for resolving resources.
In previous versions of the v3 spec, the name of an extension was required to
be a URI that dereferences to a human-readable codec specification, i.e. a URL.
That is now discouraged for new extensions, though, for backwards compatibility
with existing extensions, URLs names are still permitted.

Instead, extension names SHOULD either be registered names as specified above or URNs.
URNs (`Uniform Resource Names <https://en.wikipedia.org/wiki/Uniform_Resource_Name>`_)

> are simpler persistent identifiers assigned within defined namespaces.

Reviewer 7: TL;DR My recommendation here is to use both a UUID _and_ a tag. Having the UUID being a canonical identifier would provide a way to uniquely identify extensions should tags need to change over time due to a change in authority or circumstances.

While I think specifying tags are better than some arbitrary URI, I will also point out that this makes some assumptions and requires registration with regard to the authority portion. We are relying on either the email addresses or domain names as a registration system.

The problem I see with this is that email addresses and domain names can change over time. Josh might not always have access to josh@openmicroscopy.org. I might wonder if it would be better to contact him at his gerbi-gmb.de address at some point in the future for example.

I think we should take a hint from decentralized systems and really consider adding a UUID as a unique identifier for extensions. The UUID can help uniquely identify extensions even if the authority changes. The UUID can also be used in addition to the tag that you suggest. If someone needs to change the tag this is fine because I can still uniquely identify an extension by it's UUID. While I prefer UUIDv4, a random 128-bit sequence, other UUID variants incorporate time. For example, UUIDv7, begins with a 48-bit UNIX timestamp for example.

Support for UUID versions 6 to 8 was just added to the Python standard library:
https://github.com/python/cpython/issues/89083

Rust supports UUIDs up to version 8:
https://docs.rs/uuid/latest/uuid/

One example where UUIDs are used is the Julia packaging. A UUID is generated for each new Julia package:
https://github.com/JuliaIO/Zarr.jl/blob/8d6524678cdab7e263706f7ad37f272f8f2c3e42/Project.toml#L2

When they are registered, that information is also incorporated into the registry:
https://github.com/JuliaRegistries/General/blob/master/Z/Zarr/Package.toml

This system allows for name and repository changes. It is also robust to name collisions. I can tell two extensions apart even if they have the same name or tag. Indeed, there may be multiple incompatible implementations of the an extension with the same name. The probability of a UUID colission is extremely low.

This is unfortunately not theoretical. Implementation-specific incompatibilities have arisen with LZ4 and Zstd over time:
https://github.com/zarr-developers/numcodecs/issues/175
https://github.com/zarr-developers/zarr-python/issues/2056

Registered extensions can at least point to a source of authority if incompatibilities arise. Unregistered extensions may never have an absolute source of authority. We need a way to uniquely identify and distinguish extensions without any central authority.

> are simpler persistent identifiers assigned within defined namespaces.

Reviewer: > One example where UUIDs are used is the Julia packaging. A UUID is generated for each new Julia package:
> https://github.com/JuliaIO/Zarr.jl/blob/8d6524678cdab7e263706f7ad37f272f8f2c3e42/Project.toml#L2

something important about the julia packaging example: the _name of the package_ is not the _unique identifier_. this is an option for zarr. As I [suggested upthread](https://github.com/zarr-developers/zarr-specs/pull/330#issuecomment-2670962219), if global uniqueness is an essential design goal of this PR, then we could achieve that with scheme like `{"name": <short name>, "uuid": <long uuid>, "configuration": {...}}`.  I think this is a good way to get global uniqueness and a human-readable name.

.. _extension-guidance:

Guidance for extension authors
------------------------------

*This section is non-normative and provides assistance for the authors of
extensions, especially those who are just getting started.*

Recognizing that there are diverse considerations in choosing an extension
name, guidance is provided below based on generic scenarios. Extension authors
who are still unsure of how best to choose a name are welcome to open an issue
on the zarr-specs repository.

* **Local development**: Authors looking to define a name for local development
  purposes should prefix their extensions with ``urn:x-``. This prefix defines

>   an "experimental" name. As such an extension matures, authors might consider

Reviewer: I don't think it's easy to define "experimental" (for example, "local development" does not mean "experimental"). And I would think local development is exactly the environment where extension names would be uncontrolled. Introducing a special naming scheme for this case seems a bit weird.

> * **Local development**: Authors looking to define a name for local development
>   purposes should prefix their extensions with ``urn:x-``. This prefix defines
>   an "experimental" name. As such an extension matures, authors might consider
>   registering a new name for it. Implementations should check both for the
>   unregistered as well as the registered named.

Reviewer 4: ```suggestion
* **Local development of a new registered extension**: Authors developing a new extension intended for public
   consumption should feel free to coin a new non-prefixed "raw" name for use in development,
   with the understanding that the extension name should be registered by the time the extension is released.
```

Reviewer: This suggestion feels correct, but it complicates the definition of "unregistered names", no? Isn't a name that has not yet been registered an unregistered name? Is the definition of "unregistered name" actually "a name which the creator has no intention of ever registering"? And that seems a bit hard to predict in advance.

Reviewer 4: What do you suggest we do then?

Reviewer: _If_ we stick with the meaning of "registered names" and "unregistered names" used earlier in this PR, we must require that local development uses a URI, regardless of the developer's intent to later register. Only when the developer registers a (non-URI) name can they use that name in their implementation

Because I think this conclusion is unacceptable, we have to revise the decision to have 1 type of name for registered extensions and another very different type of name for unregistered extensions. See https://github.com/zarr-developers/zarr-specs/pull/330#issuecomment-2759392714 for an alternative proposal.

Author: Resolving comment so we can merge the current simplified state of the PR.

* **Proprietary extensions**: Authors looking to create proprietary extensions
  which are only interpretable within their own institutions are encouraged to
  take ownership of their "own" namespace, ``urn:x-company`` or

>   ``urn:x-domain.name``.

Reviewer: How do I "take ownership of a namespace"?

Reviewer 2: I would refer to other resources about URNs. tldr: You just make one up. The `urn:x-*` and `urn:example` namespaces are available for use.

Reviewer 4: `urn:x-` is deprecated. We should not recommend it. And we should definitely not expect organizations to go through IANA registration of some new top-level URN namespace just so they can create a proprietary Zarr extension!

Author: Resolving comment since namespacing is no longer a part of this PR.

> * **Proprietary extensions**: Authors looking to create proprietary extensions
>   which are only interpretable within their own institutions are encouraged to
>   take ownership of their "own" namespace, ``urn:x-company`` or
>   ``urn:x-domain.name``.

Reviewer 4: ```suggestion
* **Proprietary extensions**: Authors looking to create proprietary extensions
  for internal, non-public use are encouraged to use a Tag URI. For example
  ``tag:mycompany.com,2025-03-27:top-secret``.
```

> - The identifiers are independent of any particular resolution scheme.

These requirements are aligned well with the needs of Zarr extension developers.

An example of a Tag URI for a Zarr extension is `tag:josh@openmicroscopy.org,2025-03:experimental-new-dtype`.

.. _extension-guidance:

Guidance for extension authors
------------------------------

*This section is non-normative and provides assistance for the authors of
extensions, especially those who are just getting started.*

Recognizing that there are diverse considerations in choosing an extension
name, guidance is provided below based on generic scenarios. Extension authors
who are still unsure of how best to choose a name are welcome to open an issue
on the zarr-specs repository.

* **Local development**: Authors looking to define a name for local development
  purposes should prefix their extensions with ``urn:x-``. This prefix defines
  an "experimental" name. As such an extension matures, authors might consider
  registering a new name for it. Implementations should check both for the
  unregistered as well as the registered named.

> * **Proprietary extensions**: Authors looking to create proprietary extensions
>   for internal, non-public use are encouraged to use a Tag URI. For example
>   ``tag:mycompany.com,2025-03-27:top-secret``.

Reviewer 7: We should add some clarification that the date is not meant to imply a version number and is only meant for uniqueness of the tag. Also, we should add an example for an independent developer that does not necessarily control a domain name. I've also added a "query" component since RFC 4151 specifically points out that this is valid.

```suggestion
* **Proprietary extensions**: Authors looking to create proprietary extensions
  for internal, non-public use are encouraged to use a Tag URI. The authority and date portions of the Tag URI are meant to establish uniqueness only and do not necessarily imply contact information or a version number. For example, the names
  ``tag:mycompany.com,2025-03-27:top-secret`` or ``tag: john.smith@example.com,2038-01-19:project_foo?id=1230`` would be appropriate for private development.
```

Author: Thanks for these, guys. Resolving this comment though since this text has been stripped out.

.. _extensions_section:

Extensions
==========

This section describes how additional functionality can be defined
for Zarr datasets by the `metadata documents`_.

Extension points
----------------

Different types of extensions can exist and they can be grouped as follows:

=========== ======================= ================================================================== ================================

> node_type   extension point         metadata definition                                                list of core extensions

Reviewer 3: Fill values are kind of an extension point too, even if 3.0 does not spell it out in the extensions point section right now. The fill value metadata section has this to say 
> Extensions to the spec that define new data types must also define the JSON fill value representation.

A `string` data type needs string fill values, for example.

Reviewer 2: I tried to address this in https://github.com/zarr-developers/zarr-specs/pull/330/commits/a791ae536d57526de4c45cf7b6e6b943e601cbd0.

Reviewer 3: Nailed it

array       storage transformer     :ref:`storage-transformers <array-metadata-storage-transformers>`  :ref:`storage-transformers-list`
=========== ======================= ================================================================== ================================

New extension points may be proposed to the Zarr community through the ZEP
process. See `ZEP 0 <https://zarr.dev/zeps/active/ZEP0000.html>`_ for more information.

.. _extension-definition:

Extension definition
--------------------

In `metadata documents`_, extensions can be encoded either as objects or as
short-hand names. If using an objection definition, the member ``name``
MUST be a plain string which conforms to :ref:`extension name <extension-naming>`.
Optionally, the member ``configuration`` MAY be present but if so MUST be
an object.

For example::

    {
        "name": "<name>",        # "raw name" or URL-based name
        "configuration": { ... } # optional
    }

Instead of extension objects, short-hand names MAY be used if no
configuration metadata is required. They are equivalent to extension
objects with just a `name` key.

> If such an object is present, the field `must_understand` is implicitly set to

Reviewer 2: This part is referenced in different sections of this doc. Should it become its own (sub)section?

`True` and an object MAY explicitly set `must_understand=False` if
implementations can ignore its presence.

An implementation MUST fail to open Zarr groups or arrays if any
metadata fields are present which (a) the
implementation does not recognize and (b) are not explicitly
set to ``"must_understand": false``.

`must_understand=False` is not supported for the following extension points:
data type, chunk grid, and chunk key encoding.

.. _extension-naming:

Extension naming
----------------

The `name` field of an extension can take two forms: **raw names** and **URL-based names**.

.. _extension-naming-raw-names:

Raw names
^^^^^^^^^

Raw names are centrally registered names which can be used without prefix.

Raw names MUST be assigned within a central repository.
Raw names are unique and immutable.

> Raw names MUST be composed of lower case letters a-z, numerals 0-9, underscores, dashes, and dots.

Reviewer 2: We probably want to be a bit more restrictive, e.g. require at least on lower case letter (not just dots, dashes, underscores or numbers)

Raw name assignment is managed through the `zarr-extensions`_
Github repository, where extensions and their specification are listed.
The Zarr Steering Council or by delegation a
maintainer team reserves the right to refuse name assignment at its own
discretion.

- **Example:** ``zstd``

> - **Acceptd regex:** ``^[a-z0-9-_.]+$``

Reviewer 3: accepted

Author: Cheers.

> - **Acceptd regex:** ``^[a-z0-9-_.]+$``

Author: ```suggestion
- **Accepted regex:** ``^[a-z0-9-_.]+$``
```

.. _extension-naming-url-based-names:

URL-based names
^^^^^^^^^^^^^^^

> URL-based names delegate name registration to the domain name system (DNS).

Reviewer 2: That is not entirely right, because paths are valid. In fact, I could imagine many URL-based names would be some https://github.com/... URL.

Reviewer 2: ```suggestion
URL-based names delegate name registration to URLs on the internet.
```

Author: Very true. I'll try to say it more explicitly.

.. _extensions_section:

Extensions
==========

This section describes how additional functionality can defined

> for Zarr datasets by extended the `metadata documents`_.

Reviewer: ```suggestion
for Zarr datasets by adding fields to the `metadata documents`_.
```

Extension points
----------------

Different types of extensions can exist and they can be grouped as follows:

=========== ======================= ====================================== ================================

> level       extension               metadata                               core definitions

Reviewer: this applies to the old text, but I find these columns confusing. array is a node type, not a "level". I also don't really see the point of the "extension" column -- presumably the name of the extension point is just the relevant field in array metadata? And I'm not sure why we need the "metadata column". I think better columns would be

| node | field |
|-|-|
| array | [`data_type`](`data_type`_) |

i.e., we indicate the node type (array or group) and the name of the field, which can be a hyperlink to the actual specification for that field

=========== ======================= ====================================== ================================
array       data type               `array-metadata-data-type`_              :ref:`data-types-list`
array       chunk grid              `array-metadata-chunk-grid`_             :ref:`chunk-grid-list`
array       chunk key encoding      `array-metadata-chunk-key-encoding`_     :ref:`chunk-key-encoding-list`
array       codecs                  `array-metadata-codecs`_                 :ref:`codecs-list`
array       storage transformer     `array-metadata-storage-transformers`_   :ref:`storage-transformers-list`
=========== ======================= ====================================== ================================

> If such extension points are used by groups or arrays, they are required.

Reviewer: old text but what does this sentence mean?

Author: This is perhaps the (terse) explanation of your point in https://github.com/zarr-developers/zarr-specs/pull/330/files#r1962198166 -- i.e. use of these extension points falls under `must_understand=True`. I'll look into unifying.

New extension points may be proposed to the Zarr community through the ZEP
process. See `ZEP 0 <https://zarr.dev/zeps/active/ZEP0000.html>`_ for more information.

Extension definition
--------------------

> Extensions are defined in `metadata documents`_ either as objects or as

Reviewer: I find the statement "extensions are defined in metadata documents" confusing. Aren't extensions defined in specification documents?

Author: I think this comes back to our different understand of the terminologies, so deferring for https://github.com/zarr-developers/zarr-specs/pull/330/files#r1962039797

short-hand names. If using an objection definition, the following pattern
MUST be followed::

    {
        "name": "<name>",
        "configuration": { ... } # optional
    }

If such an object is present, the field `must_understand` is implicitly set to
`True` and an object can explicitly set `must_understand=False` if
implementations can ignore its presence, following the current guidelines in

> the v3 specification.

Reviewer: Presumably it's not possible to have a data type with `must_understand=False`? I think we are confusing "non-core data type" with "extra field in `zarr.json`". these are two very different forms of extension, and they should be discussed separately.

Author: Good point. I also can't see a way for there to be a sensible data type, chunk grid or key encoding with `must_understand=False`. It's a bit far-fetched, but e.g. a logging codec could potentially be optional, as long as the semantics are "then act as if it doesn't exist in the pipeline". (Ignoring storage transformers). Certainly other future extension points might also be able to make use of the must_understand flag.

This will need clarification.

Instead of extension objects, short-hand names may continue to be used if no
configuration metadata is required. They would be equivalent to extension
objects with just a `name` key.

Extension naming
----------------

> The identifier used in the `name` field of the extension definition can follow one of two forms:

Reviewer: ```suggestion
There are two types of extension identifiers:
```

> 1. **Raw names** MUST be assigned within a central repository and follow the

Reviewer: I would open by saying what a raw name is

> 1. **Raw names** MUST be assigned within a central repository and follow the
>    compatibility and versioning v3 `stability policy`_.

Reviewer: ```suggestion
1. **Raw names** MUST be assigned within a central repository and follow the
   compatibility and versioning v3 `stability policy`_.
```

how can a string follow a compatibility or versioning policy?

>    compatibility and versioning v3 `stability policy`_.

Reviewer: how can a string follow a compatibility or versioning policy?

   The name assignment is managed through the `zarr-extensions`_
   Github repository, where each extension is
   listed and either contains a spec document or links to a spec document.

>    Names are never unassigned or reassigned. The Zarr Steering Council or by delegation a

Reviewer: ```suggestion
   Raw names are unique and immutable. The Zarr Steering Council or by delegation a
```

   maintainer team reserves the right to refuse name assignment at its own
   discretion.

   - **Example:** ``zstd``

>    - **Acceptd regex:** ``^[a-z0-9-_.]+$``

Reviewer: ```suggestion
   - **Acceptd regex:** ``^[a-z0-9-_.]+$``
```
Maybe I'm alone here but I don't find regex useful. If that regex can be translated to english, this would be better.

Reviewer 2: I think regexes are nice because they are unambiguous.

> 2. **URI-based names** can be used by anyone without further coordination

Reviewer: ```suggestion
2. **URI-based names** can be used by any extension without further coordination
```

>    though the assumption is that users reasonably "own" the URI. The URL SHOULD

Reviewer: what does it mean to "reasonably own" a URI?

>    resolve to a human-readable explanation of the extension, but

Reviewer: ```suggestion
   resolve to a JSON schema definition of the extension metadata.
```

Author: I'll add as even better than just human readable.

Reviewer: proposing JSON schema here because that's something implementations can actually use

>    implementations SHOULD NOT attempt to resolve the URL during processing.

Reviewer: ```suggestion
```

Reviewer: I don't think we should tell implementations what to do here

Author: I'll change to say there's no expectation of such.

>    There are no guarantees in terms of versioning or compatibility. However,

Reviewer: this section is about extension naming, but if we are talking about versioning and backwards compatibility then it seems like we are actually talking about extensions

   preserving backwards-compatibility is strongly encouraged. See the
   [versioning section](#Versioning-and-spec-evolution) below.

   - **Example:** ``https://example.com/zarr3/consolidated-metadata``

>    - **Accepted regex:** ``^https?:\/\/[^/?#]+[^?#]*$``

Reviewer: why is https required?

Reviewer 2: It is `http` or `https`. The `?` in the regex makes the `s` optional.

Reviewer: so then why is http or https required?

Extension example
-----------------

> The following example represents an Array showing many of the proposed changes

Reviewer: ```suggestion
The following example of array metadata demonstrates these extension naming schemes.
```

described above::

    {
        "zarr_format": 3,
        "data_type": "https://example.com/zarr/string", // URI-based name, short-hand name
        "chunk_key_encoding": {
            "name": "default", // core
            "configuration": { "separator": "." }
        },
        "codecs": [
            {
                "name": "https://numcodecs.dev/vlen-utf8" // URI-based name
            },
            {
                "name": "zstd", // raw name
                "configuration": { ... }
            }
        ],
        "chunk_grid": {
            "name": "regular", // core
            "configuration": { "chunk_shape": [ 32 ] }
        },
        "shape": [ 128 ],
        "dimension_names": [ "x" ],
        "attributes": { ... },
        "storage_transformers": []
    }

Extension specifications
------------------------

> There is no strict requirement for extensions to have a formal specification.
> However, for adoption in the community it is STRONGLY RECOMMENDED to write a
> specification.

Reviewer: ```suggestion
Extensions SHOULD have a published specification. A published specification facilitates multiple implementations of an extension.
```

For extensions with raw names, the `zarr-extensions`_ repository

> SHOULD either contain the specification directly or link to the official location.

Reviewer: ```suggestion
SHOULD either contain the specification or link to it.
```

> For extensions with URI-based names, it is RECOMMENDED to publish the specification
> under the URI of the extension. Additionally, URI-based extensions MAY also register

Reviewer: ```suggestion
For extensions with URI-based names, it is RECOMMENDED that the URI resolve to a specification of the extension. Additionally, URI-based extensions MAY also register
```

are documented in this section, grouped by the specification status and ordered
by time.

3.1
---

- Clarification of extensions. `PR #330
  <https://github.com/zarr-developers/zarr-specs/pull/330/>`_. With this change,

>   it is now possible to register new names or even use URLs for extensions.

Reviewer 4: ```suggestion
  it is now possible to register new names or used namespace-prefixed names for extensions.
```
